File: JS/mergesort.js
//merges two already sorted arrays into one array
function merge(arr1, arr2){
      let i = 0, j = 0, mergedArr = [];
      while (i < arr1.length && j < arr2.length){
            if (arr1[i] > arr2[j]) mergedArr.push(arr2[j++]);
            else mergedArr.push(arr1[i++]);
      }
      while (i < arr1.length){
            mergedArr.push(arr1[i++]);
      }
      while (j < arr2.length){
            mergedArr.push(arr2[j++]);
      }
  return mergedArr;
}

//mergesort function
function mergeSort(array){
    //check array length, if its 1 then its already sorted
    if(array.length==1){
        return array;
    }
    //split the array into two halves
    let m = Math.floor(array.length/2);
    //recursively sort the two new arrays
    let arr1 = mergeSort(array.slice(0,m));
    let arr2 = mergeSort(array.slice(m));
    //merge the sorted arrays back together
    return merge(arr1, arr2);
}

File: Python/merge_sort.py
# this program implements a mergesort algorithm
# merge sort divides the list into equal halves then combines them in a sorted manner

def merge_sort(list):

    if(len(list) <= 1):
        return list #list is already sorted

    mid = len(list)//2 #defines the midpoint and slices the list into two halves
    left = list[:mid]
    right = list[mid:]

    left = merge_sort(left)
    right = merge_sort(right) #recursively mergesort the left and right lists

    return merge(left,right)


def merge(left, right):

    result = []

    while(len(left) != 0 and len(right) != 0): #merge the two lists together
        if(left[0] < right[0]):
            result.append(left[0])
            left.remove(left[0])
        else:
            result.append(right[0])
            right.remove(right[0])
    
    if(len(left) == 0):
        result = result + right
    else:
        result = result + left
    
    return result

#example list for testing
list = [64, 34, 25, 12, 22, 11, 90]
print(merge_sort(list))